kernel/cls/cls_macros/src/lib.rs

```diff
@@ -87,22 +148,114 @@ pub fn cpu_local(args: TokenStream, input: TokenStream) -> TokenStream {
+            #[cfg(target_arch = "x86_64")]
+            {
+                unsafe {
+                    ::core::arch::asm!(
+                        "rdgsbase {}",
+                        out(reg) ptr,
+                        options(nomem, preserves_flags, nostack),
+                    )
+                };
+            }
+            #[cfg(target_arch = "aarch64")]
+            {
+                unsafe {
+                    ::core::arch::asm!(
+                        "mrs {}, tpidr_el1",
+                        out(reg) ptr,
+                        options(nomem, preserves_flags, nostack),
+                    )
+                };
+            }
+            ptr += #offset;
+            unsafe { &mut*(ptr as *mut #ty) }
+        }
+    };
+
+    let cls_dep_functions = if cls_dependency {
+        let guarded_functions = quote! {
+            #[inline]
+            pub fn replace_guarded<G>(&self, mut value: #ty, guard: &G) -> #ty
+            where
+                G: ::cls::Guard,
+            {
+                let rref = #ref_expr;
+                ::core::mem::swap(rref, &mut value);
+                value
+            }
+
+            #[inline]
+            pub fn set_guarded<G>(&self, mut value: #ty, guard: &G)
+            where
+                G: ::cls::Guard,
+            {
+                self.replace_guarded(value, guard);
             }
         };
 
-    let x64_width_modifier = format!("{x64_asm_width} ptr ");
-    let x64_cls_location = format!("gs:[{offset}]");
+        if let Some(guard_type) = stores_guard {
+            quote! {
+                #guarded_functions
+
+                #[inline]
+                pub fn replace(&self, guard: #guard_type) -> #ty {
+                    // Check that the guard type matches the type of the static.
+                    trait TyEq {}
+                    impl<T> TyEq for (T, T) {}
+                    fn ty_eq<A, B>()
+                    where
+                        (A, B): TyEq
+                    {}
+                    ty_eq::<::core::option::Option<#guard_type>, #ty>();
+
+                    // Check that the guard type implements cls::Guard.
+                    fn implements_guard_trait<T>()
+                    where
+                        T: ::cls::Guard
+                    {}
+                    implements_guard_trait::<#guard_type>();
+
+                    let rref = #ref_expr;
+                    let mut guard = Some(guard);
+                    ::core::mem::swap(rref, &mut guard);
+
+                    guard
+                }
+
+                #[inline]
+                pub fn set(&self, mut guard: #guard_type) {
+                    self.replace(guard);
+                }
+            }
+        } else {
+            quote! {
+                #guarded_functions
+
+                #[inline]
+                pub fn replace(&self, value: #ty) -> #ty {
+                    // TODO: Should this ever be `disable_interrupts` instead?
```

Comment:
not by default, no, but IIRC we had discussed a prior idea of using the `DeadlockPrevention` trait here to allow the user/definer of the CLS variable to specify

Instead of doing that, we can just use your new `Guard` trait (which could be further abstracted as an "atomicity-ensurer" trait or something) for that purpose, as using a guard for atomicity is more accurate for what we want to achieve here than preventing deadlock.

Reply:
We can't define default type for functions so I think for now it's best not to because it would just add more complexity to the API for no apparent gain.

Reply:
Yeah, no problem. We could define it on the struct perhaps, but no need.